### PR TITLE
[MODEL] Mistral Nemo 12B Support

### PR DESCRIPTION
**Pull request by casper-hansen**
#### Context
What is the purpose of this PR? Is it to
- [x] add a new feature
- [ ] fix a bug
- [ ] update tests and/or documentation
- [ ] other (please add here)

Issue: #1716

#### Changelog
What are the changes made in this PR?

* Added Mistral's "v3" tokenizer called Tekken (tiktoken based)
* Added support for specifying `head_dim` (default=None for backward compatibility) in the MistralModel for loading the model weights with correct shapes

#### Test plan
Please make sure to do each of the following if applicable to your PR. If you're unsure about any one of these just ask and we will happily help. We also have a [contributing page](https://github.com/pytorch/torchtune/blob/main/CONTRIBUTING.md) for some guidance on contributing.

- [ ] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [ ] add [unit tests](https://github.com/pytorch/torchtune/tree/main/tests/torchtune) for any new functionality
- [ ] update [docstrings](https://github.com/pytorch/torchtune/tree/main/docs/source) for any new or updated methods or classes
- [ ] run unit tests via `pytest tests`
- [ ] run recipe tests via `pytest tests -m integration_test`
- [ ] manually run any new or modified recipes with sufficient proof of correctness
- [ ] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

#### UX
If your function changed a public API, please add a dummy example of what the user experience will look like when calling it.
Here is a [docstring example](https://github.com/pytorch/torchtune/blob/6a7951f1cdd0b56a9746ef5935106989415f50e3/torchtune/modules/vision_transformer.py#L285)
and a [tutorial example](https://pytorch.org/torchtune/main/tutorials/qat_finetune.html#applying-qat-to-llama3-models)

- [x] I did not change any public API
- [ ] I have added an example to docs or docstrings

### CODE REVIEW

**Comment by casper-hansen** [comment]
Before I go onto add tests, align the config to other configs, implement LoRA etc., I would like to know if this PR is likely to be accepted to support the Mistral Nemo model if those things are implemented. CC @joecummings

**Comment by pbontrager** [comment]
Thank you for this contribution! Could you add some context on how you verified this model and link to any gist where you compared the model outputs to HF? Ideally we could see that this tokenizer matches the official implementation exactly for some dummy inputs and that generation with the model weights (using the generate.py recipe) is coherent and that finetuning with the model seems to work.

**Comment by casper-hansen** [comment]
@pbontrager I appreciate your review. I will not have time to go through more testing. You are free to take this code, modify it, etc. I tested the training and compared it to Mistral 7B with the same config and the loss was within the same ranges and seemed normal.
